### ISSUE TITLE
The data in a table created by a select statement is different from the result of that select.

### ISSUE THREAD

**Issue by MikhailBurdukov** [ClickHouse/ClickHouse#59930]
**How to reproduce**
Have reproduced on:
clickhouse-23.12, 24.1

Haven't tried older versions. 

Preparation:
```
CREATE DATABASE test_db ON CLUSTER '{cluster}'

CREATE TABLE test_db.t_table_select ON CLUSTER `{cluster}`
(
    `id` UInt32
)
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}')
ORDER BY id
SETTINGS index_granularity = 8192

CREATE TABLE test_db.d_table_select ON CLUSTER `{cluster}`
(
    `id` UInt32
)
ENGINE = Distributed('{cluster}', 'test_db', 't_table_select', id)

INSERT INTO test_db.d_table_select (id) SELECT number FROM numbers(30)


CREATE TEMPORARY TABLE t_test AS 
SELECT
      a.id, b.id   
FROM test_db.d_table_select AS a
GLOBAL LEFT JOIN
(
    SELECT id
    FROM test_db.d_table_select AS b
    WHERE (b.id % 10) = 0
) AS b ON b.id = a.id
SETTINGS join_use_nulls = 1

```

Then the data in the `t_test` table and from the inner select will be different:
The select output: 
```
SELECT
    a.id,
    b.id
FROM test_db.d_table_select AS a
GLOBAL LEFT JOIN
(
    SELECT id
    FROM test_db.d_table_select AS b
    WHERE (b.id % 10) = 0
) AS b ON b.id = a.id
SETTINGS join_use_nulls = 1

Query id: b8f09d77-058c-47ea-938f-3e0c541fe6a9

┌─id─┬─b.id─┐
│  0 │    0 │
│  1 │ ᴺᵁᴸᴸ │
│  2 │ ᴺᵁᴸᴸ │
│  3 │ ᴺᵁᴸᴸ │
│  4 │ ᴺᵁᴸᴸ │
│  5 │ ᴺᵁᴸᴸ │
│  6 │ ᴺᵁᴸᴸ │
│  7 │ ᴺᵁᴸᴸ │
│  8 │ ᴺᵁᴸᴸ │
│  9 │ ᴺᵁᴸᴸ │
│ 10 │   10 │
│ 11 │ ᴺᵁᴸᴸ │
│ 12 │ ᴺᵁᴸᴸ │
│ 13 │ ᴺᵁᴸᴸ │
│ 14 │ ᴺᵁᴸᴸ │
│ 15 │ ᴺᵁᴸᴸ │
│ 16 │ ᴺᵁᴸᴸ │
│ 17 │ ᴺᵁᴸᴸ │
│ 18 │ ᴺᵁᴸᴸ │
│ 19 │ ᴺᵁᴸᴸ │
│ 20 │   20 │
│ 21 │ ᴺᵁᴸᴸ │
│ 22 │ ᴺᵁᴸᴸ │
│ 23 │ ᴺᵁᴸᴸ │
│ 24 │ ᴺᵁᴸᴸ │
│ 25 │ ᴺᵁᴸᴸ │
│ 26 │ ᴺᵁᴸᴸ │
│ 27 │ ᴺᵁᴸᴸ │
│ 28 │ ᴺᵁᴸᴸ │
│ 29 │ ᴺᵁᴸᴸ │
└────┴──────┘
```
The data in t_test table:
```
SELECT *
FROM t_test

Query id: 64c3fd04-f312-4c7c-8372-d81ed2f5e5bc

┌─id─┬─b.id─┐
│  0 │ ᴺᵁᴸᴸ │
│  1 │ ᴺᵁᴸᴸ │
│  2 │ ᴺᵁᴸᴸ │
│  3 │ ᴺᵁᴸᴸ │
│  4 │ ᴺᵁᴸᴸ │
│  5 │ ᴺᵁᴸᴸ │
│  6 │ ᴺᵁᴸᴸ │
│  7 │ ᴺᵁᴸᴸ │
│  8 │ ᴺᵁᴸᴸ │
│  9 │ ᴺᵁᴸᴸ │
│ 10 │ ᴺᵁᴸᴸ │
│ 11 │ ᴺᵁᴸᴸ │
│ 12 │ ᴺᵁᴸᴸ │
│ 13 │ ᴺᵁᴸᴸ │
│ 14 │ ᴺᵁᴸᴸ │
│ 15 │ ᴺᵁᴸᴸ │
│ 16 │ ᴺᵁᴸᴸ │
│ 17 │ ᴺᵁᴸᴸ │
│ 18 │ ᴺᵁᴸᴸ │
│ 19 │ ᴺᵁᴸᴸ │
│ 20 │ ᴺᵁᴸᴸ │
│ 21 │ ᴺᵁᴸᴸ │
│ 22 │ ᴺᵁᴸᴸ │
│ 23 │ ᴺᵁᴸᴸ │
│ 24 │ ᴺᵁᴸᴸ │
│ 25 │ ᴺᵁᴸᴸ │
│ 26 │ ᴺᵁᴸᴸ │
│ 27 │ ᴺᵁᴸᴸ │
│ 28 │ ᴺᵁᴸᴸ │
│ 29 │ ᴺᵁᴸᴸ │
└────┴──────┘
```

**Expected behavior**
The data in `t_test` table must be the same as in the select query.

**Error message and/or stacktrace**
Logs seems fine.

**Comment by antaljanosbenjamin** [comment]
I tried to narrow the scope of the bug, thus I concluded the following things:
1. `SETTINGS join_use_nulls = 1` is not necessary. As a result instead of nulls we will see 0s, but the bug remains.
2. The `SELECT` can be trivial, e.g.: `SELECT 10 AS id`.
3. It is important to have `GLOBAL` in the query and the distributed table on the left side of the join. Otherwise the bug doesn't reproduce.

**Comment by antaljanosbenjamin** [comment]
I checked and it is not present in 23.11 (40cfda2b6df43540583123bc8c3639d3672cbcff).

**Comment by antaljanosbenjamin** [comment]
It is present in the first commit on the 23.12 branch (53e1194b8e83524f9f8f48b03f777ec229147053).

**Comment by antaljanosbenjamin** [comment]
The issue was introduced in #57855.

**Comment by antaljanosbenjamin** [comment]
Workaround is to use new analyzer when the temporary table is created:
```
ip-10-2-3-206.eu-west-1.compute.internal :) CREATE DATABASE test_db ON CLUSTER '{cluster}';

CREATE TABLE test_db.t_table_select ON CLUSTER `{cluster}`
(
    `id` UInt32
)
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}')
ORDER BY id
SETTINGS index_granularity = 8192;

CREATE TABLE test_db.d_table_select ON CLUSTER `{cluster}`
(
    `id` UInt32
)
ENGINE = Distributed('{cluster}', 'test_db', 't_table_select', id);

INSERT INTO test_db.d_table_select (id) SELECT number FROM numbers(30);


CREATE TEMPORARY TABLE t_test AS 
SELECT
      a.id, b.id   
FROM test_db.d_table_select AS a
GLOBAL LEFT JOIN
(
    SELECT id
    FROM test_db.d_table_select AS b
    WHERE (b.id % 10) = 0
) AS b ON b.id = a.id
SETTINGS join_use_nulls = 1 SETTINGS allow_experimental_analyzer=1;

CREATE DATABASE test_db ON CLUSTER `{cluster}`

Query id: 3792de00-91dd-4118-a049-2af7e687ce9c

┌─host──────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ localhost │ 9000 │      0 │       │                   0 │                0 │
└───────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

1 row in set. Elapsed: 0.104 sec. 


CREATE TABLE test_db.t_table_select ON CLUSTER `{cluster}`
(
    `id` UInt32
)
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}')
ORDER BY id
SETTINGS index_granularity = 8192

Query id: 5eaa5a1a-b120-4278-8af1-9f6113716589

┌─host──────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ localhost │ 9000 │      0 │       │                   0 │                0 │
└───────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

1 row in set. Elapsed: 0.103 sec. 


CREATE TABLE test_db.d_table_select ON CLUSTER `{cluster}`
(
    `id` UInt32
)
ENGINE = Distributed('{cluster}', 'test_db', 't_table_select', id)

Query id: 29f8dc6e-4a89-42f4-82da-2c9bf8f5c486

┌─host──────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
│ localhost │ 9000 │      0 │       │                   0 │                0 │
└───────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

1 row in set. Elapsed: 0.103 sec. 


INSERT INTO test_db.d_table_select (id) SELECT number
FROM numbers(30)

Query id: d3c06171-29a2-4f23-a3c8-3b03a86a2096

Ok.

0 rows in set. Elapsed: 0.003 sec. 


CREATE TEMPORARY TABLE t_test AS
SELECT
    a.id,
    b.id
FROM test_db.d_table_select AS a
GLOBAL LEFT JOIN
(
    SELECT id
    FROM test_db.d_table_select AS b
    WHERE (b.id % 10) = 0
) AS b ON b.id = a.id
SETTINGS join_use_nulls = 1
SETTINGS allow_experimental_analyzer = 1

Query id: 41d0a1c2-878b-40f5-b79c-8a66503f3f16

Ok.

0 rows in set. Elapsed: 0.005 sec. 

ip-10-2-3-206.eu-west-1.compute.internal :) SELECT * FROM t_test;

SELECT *
FROM t_test

Query id: ef178e74-e9bc-42d8-bc6a-df136413c51d

┌─a.id─┬─b.id─┐
│    0 │    0 │
│    1 │ ᴺᵁᴸᴸ │
│    2 │ ᴺᵁᴸᴸ │
│    3 │ ᴺᵁᴸᴸ │
│    4 │ ᴺᵁᴸᴸ │
│    5 │ ᴺᵁᴸᴸ │
│    6 │ ᴺᵁᴸᴸ │
│    7 │ ᴺᵁᴸᴸ │
│    8 │ ᴺᵁᴸᴸ │
│    9 │ ᴺᵁᴸᴸ │
│   10 │   10 │
│   11 │ ᴺᵁᴸᴸ │
│   12 │ ᴺᵁᴸᴸ │
│   13 │ ᴺᵁᴸᴸ │
│   14 │ ᴺᵁᴸᴸ │
│   15 │ ᴺᵁᴸᴸ │
│   16 │ ᴺᵁᴸᴸ │
│   17 │ ᴺᵁᴸᴸ │
│   18 │ ᴺᵁᴸᴸ │
│   19 │ ᴺᵁᴸᴸ │
│   20 │   20 │
│   21 │ ᴺᵁᴸᴸ │
│   22 │ ᴺᵁᴸᴸ │
│   23 │ ᴺᵁᴸᴸ │
│   24 │ ᴺᵁᴸᴸ │
│   25 │ ᴺᵁᴸᴸ │
│   26 │ ᴺᵁᴸᴸ │
│   27 │ ᴺᵁᴸᴸ │
│   28 │ ᴺᵁᴸᴸ │
│   29 │ ᴺᵁᴸᴸ │
└──────┴──────┘

30 rows in set. Elapsed: 0.001 sec.
```

**Comment by SaltTan** [comment]
Fixed in 24.2.2.71

**Comment by antaljanosbenjamin** [comment]
Yes, it was fixed by https://github.com/ClickHouse/ClickHouse/pull/60464.
